Index p_bus out of size in audio server

<!-- Please search existing issues for potential duplicates before filing yours:
https://github.com/godotengine/godot/issues?q=is%3Aissue
-->

Copypasta from https://github.com/godotengine/godot/issues/13274#issuecomment-435283146

**Godot version:**
<!-- Specify commit hash if non-official. -->
The Godot I use is 3.0.6, https://github.com/mjtorn/godot/commits/branch.3.0.6-stable.fontshadow.videofix.spine.crashfix2

**OS/device including version:**
<!-- Specify GPU model and drivers if graphics-related. -->
Linux, Audio-GD R2R11 (Combo384 Amanero) and/or whatever the amdgpu (Ellesmere [Radeon RX 470/480]) might use for HDMI audio.

**Issue description:**
<!-- What happened, and what was expected. -->
```
ERROR: thread_get_channel_mix_buffer: Index p_bus=107 out of size (buses.size()=4)
   At: servers/audio_server.cpp:432.
handle_crash: Program crashed with signal 11
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[1] /lib/x86_64-linux-gnu/libc.so.6(+0x33060) [0x7f89c608e060] (??:0)
[2] /home/mjt/src/git_checkouts/godot/bin/godot.x11.opt.tools.64(+0x1016ce1) [0x55d9b2712ce1] (??:0)
[3] /home/mjt/src/git_checkouts/godot/bin/godot.x11.opt.tools.64(+0x15cd0c7) [0x55d9b2cc90c7] (??:0)
[4] /home/mjt/src/git_checkouts/godot/bin/godot.x11.opt.tools.64(+0x15ce1cf) [0x55d9b2cca1cf] (??:0)
[5] /home/mjt/src/git_checkouts/godot/bin/godot.x11.opt.tools.64(+0x7adc5e) [0x55d9b1ea9c5e] (??:0)
[6] /home/mjt/src/git_checkouts/godot/bin/godot.x11.opt.tools.64(+0x7bc111) [0x55d9b1eb8111] (??:0)
[7] /lib/x86_64-linux-gnu/libpthread.so.0(+0x7494) [0x7f89c6ea2494] (??:0)
[8] /lib/x86_64-linux-gnu/libc.so.6(clone+0x3f) [0x7f89c6143acf] (??:0)
-- END OF BACKTRACE --
```

The expected behavior is not to segfault :D

**Steps to reproduce:**
Our game's first scene is a logo animation as OGV.

Hit `F5` in the editor repeatedly until it crashes.

**Minimal reproduction project:**
<!-- Recommended as it greatly speeds up debugging. Drag and drop a zip archive to upload it. -->
Sadly there is none.

-------

Quoting the rest of the referenced issue:

The video is an OGV which works thanks to 6dc20adadd721bfc31a6b761eb6224975938dbf4 (but I run this in the editor and iirc this always worked in the editor anyway)

The `VideoPlayer` is configured to send audio straight to the `Master` bus, as per defaults.

I can hit `F5` and it's _fairly_ consistent, but not quite. Exported games segfault.

The `Master` bus, when it fails, seems to consistently give `p_bus=107`.

Changing the bus to `Music` gives the same error but with `p_bus=80`. Changing to `Speech` gives `p_bus=121`. These seem likewise consistent in their faulty values.

Changing to `SFX` gave `p_bus=199856992` and `p_bus=2139047568` and `p_bus=1505727408`. I did not investigate further, to keep my blood pressure from rising further. It seems very inconsistent in its faulty values (not my blood pressure but the `p_bus` value).

The `default_bus_layout.tres` file we use is this:

```
[gd_resource type="AudioBusLayout" format=2]

[resource]

bus/0/name = "Master"
bus/0/solo = false
bus/0/mute = false
bus/0/bypass_fx = false
bus/0/volume_db = 0.0
bus/0/send = ""
bus/1/name = "Music"
bus/1/solo = false
bus/1/mute = false
bus/1/bypass_fx = false
bus/1/volume_db = 0.0
bus/1/send = "Master"
bus/2/name = "Speech"
bus/2/solo = false
bus/2/mute = false
bus/2/bypass_fx = false
bus/2/volume_db = 0.0
bus/2/send = "Master"
bus/3/name = "SFX"
bus/3/solo = false
bus/3/mute = false
bus/3/bypass_fx = false
bus/3/volume_db = 0.0
bus/3/send = "Master"
```

It may be a placebo effect, but I powered up my USB DAC and the issue hasn't appeared, but with bugs that are only _fairly_ consistent, we really can only speculate. I don't trust my memory on this, but with an exported game, I think I had the DAC on, and restarted it, but let's not count on that.

What is this? Some kind of crap in memory, maybe near the last bus, that causes the first three buses to have consistently same invalid values but the `SFX` one not?

Might it be related to drivers? If so, I do run Linux. It could be just another year-of-the-desktop moment, but it would be nice if this worked.

Thanks!

Getting the video scene to run was tough, it seems Escoria is very much incompatible with Godot 3.1. By some creative commenting-out of code and replacing functions' bodies with `return false`, I witnessed a crash through the editor:

```
ERROR: thread_get_channel_mix_buffer: Index p_bus=-592322416 out of size (buses.size()=4)
   At: servers/audio_server.cpp:465.
handle_crash: Program crashed with signal 11
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[1] /lib/x86_64-linux-gnu/libc.so.6(+0x33060) [0x7f6af10f0060] (??:0)
[2] /home/mjt/src/git_checkouts/godot-3.1/bin/godot.x11.opt.tools.64(+0x1278639) [0x556ad7dd5639] (??:0)
[3] /home/mjt/src/git_checkouts/godot-3.1/bin/godot.x11.opt.tools.64(+0x21ea0f3) [0x556ad8d470f3] (??:0)
[4] /home/mjt/src/git_checkouts/godot-3.1/bin/godot.x11.opt.tools.64(+0xc3203f) [0x556ad778f03f] (??:0)
[5] /home/mjt/src/git_checkouts/godot-3.1/bin/godot.x11.opt.tools.64(+0xb63d46) [0x556ad76c0d46] (??:0)
[6] /lib/x86_64-linux-gnu/libpthread.so.0(+0x7494) [0x7f6af196b494] (??:0)
[7] /lib/x86_64-linux-gnu/libc.so.6(clone+0x3f) [0x7f6af11a5acf] (??:0)
-- END OF BACKTRACE --
```

`tools=no target=debug` did not compile:

```
/tmp/ccwwZv3b.ltrans0.ltrans.o:(.data.rel.local+0x7dd0): undefined reference to `.L715'
```

`tools=no target=release_debug` is linking forever, I wonder if the compile will even succeed.

--- a/scene/gui/video_player.cpp
+++ b/scene/gui/video_player.cpp
@@ -90,53 +90,32 @@ void VideoPlayer::_mix_audio() {
 
 	AudioFrame vol = AudioFrame(volume, volume);
 
-	// Copy to server's audio buffer
-	switch (AudioServer::get_singleton()->get_speaker_mode()) {
+	int cc = AudioServer::get_singleton()->get_channel_count();
 
-		case AudioServer::SPEAKER_MODE_STEREO: {
-			AudioFrame *target = AudioServer::get_singleton()->thread_get_channel_mix_buffer(bus_index, 0);
+	if (cc == 1) {
+		AudioFrame *target = AudioServer::get_singleton()->thread_get_channel_mix_buffer(bus_index, 0);
+		ERR_FAIL_COND(!target);
 
-			for (int j = 0; j < buffer_size; j++) {
+		for (int j = 0; j < buffer_size; j++) {
 
-				target[j] += buffer[j] * vol;
+			target[j] += buffer[j] * vol;
+		}
+
+	} else {
+		AudioFrame *targets[4];
+
+		for (int k = 0; k < cc; k++) {
+			targets[k] = AudioServer::get_singleton()->thread_get_channel_mix_buffer(bus_index, k);
+			ERR_FAIL_COND(!targets[k]);
+		}
+
+		for (int j = 0; j < buffer_size; j++) {
+
+			AudioFrame frame = buffer[j] * vol;
+			for (int k = 0; k < cc; k++) {
+				targets[k][j] += frame;
 			}
-
-		} break;
-		case AudioServer::SPEAKER_SURROUND_31: {
-
-			// FIXME: Implement
-		} break;
-		case AudioServer::SPEAKER_SURROUND_51: {
-
-			AudioFrame *targets[2] = {
-				AudioServer::get_singleton()->thread_get_channel_mix_buffer(bus_index, 1),
-				AudioServer::get_singleton()->thread_get_channel_mix_buffer(bus_index, 2),
-			};
-
-			for (int j = 0; j < buffer_size; j++) {
-
-				AudioFrame frame = buffer[j] * vol;
-				targets[0][j] = frame;
-				targets[1][j] = frame;
-			}
-		} break;
-		case AudioServer::SPEAKER_SURROUND_71: {
-
-			AudioFrame *targets[3] = {
-				AudioServer::get_singleton()->thread_get_channel_mix_buffer(bus_index, 1),
-				AudioServer::get_singleton()->thread_get_channel_mix_buffer(bus_index, 2),
-				AudioServer::get_singleton()->thread_get_channel_mix_buffer(bus_index, 3)
-			};
-
-			for (int j = 0; j < buffer_size; j++) {
-
-				AudioFrame frame = buffer[j] * vol;
-				targets[0][j] += frame;
-				targets[1][j] += frame;
-				targets[2][j] += frame;
-			}
-
-		} break;
+		}
 	}
 }
 

